sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -148,80 +158,111 @@ def fit(self, X, y):
+                # predictions of the initial model (before the first tree)
+
+                if self.scoring == 'loss':
+                    # we're going to compute scoring w.r.t the loss. As losses
+                    # take raw predictions as input (unlike the scorers), we
+                    # can optimize a bit and avoid repeating computing the
+                    # predictions of the previous trees. We'll re-use
+                    # raw_predictions (as it's needed for training anyway) for
+                    # evaluating the training loss, and create
+                    # raw_predictions_val for storing the raw predictions of
+                    # the validation data.
+
+                    if self._use_validation_data:
+                        raw_predictions_val = np.zeros(
+                            shape=(self.n_trees_per_iteration_,
+                                   X_binned_val.shape[0]),
+                            dtype=self._baseline_prediction.dtype
+                        )
+
+                        raw_predictions_val += self._baseline_prediction
+
+                    self._check_early_stopping_loss(raw_predictions, y_train,
+                                                    raw_predictions_val, y_val)
+                else:
+                    self.scorer_ = check_scoring(self, self.scoring)
+                    # scorer_ is a callable with signature (est, X, y) and
+                    # calls est.predict() or est.predict_proba() depending on
+                    # its nature.
+                    # Unfortunately, each call to scorer_() will compute
+                    # the predictions of all the trees. So we use a subset of
+                    # the training set to compute train scores.
+                    subsample_size = 10000  # should we expose this parameter?
+                    indices = np.arange(X_binned_train.shape[0])
+                    self._indices = indices
+                    if X_binned_train.shape[0] > subsample_size:
+                        # TODO: not critical but stratify using resample()
+                        indices = rng.choice(indices, subsample_size,
+                                             replace=False)
+                    X_binned_small_train = X_binned_train[indices]
+                    y_small_train = y_train[indices]
+                    # Predicting is faster on C-contiguous arrays.
+                    X_binned_small_train = np.ascontiguousarray(
+                        X_binned_small_train)
+
+                    self._check_early_stopping_scorer(
+                        X_binned_small_train, y_small_train,
+                        X_binned_val, y_val,
+                    )
+            begin_at_stage = 0
+
+        else:
+            # Convert array attributes to lists
+            self._tolist_state()
+
+            # Compute raw predictions
+            raw_predictions = self._raw_predict(X_binned_train)
+            if hasattr(self, '_indices'):
```

Comment:
hmm is there a way to avoid this? We try to avoid saving these kind of stateful attributes as much as possible (the random state is an exception)

Reply:
Yes. It can be done by:
* making sure that the conditions are met: 
```python
if self.do_early_stopping_ and self.scoring != 'loss':
```
* computing the indices again:
```python
subsample_size = 10000  # should we expose this parameter?
indices = np.arange(X_binned_train.shape[0])
self._indices = indices
if X_binned_train.shape[0] > subsample_size:
    # TODO: not critical but stratify using resample()
    indices = rng.choice(indices, subsample_size,  replace=False)
```
Since we already recompute several things (binning step), we can also do it for the indices. Should I replace `choice` with `sklearn.utils.resample`?

Reply:
Ok maybe use a small helper to compute the indices then.

>Should I replace choice with sklearn.utils.resample?

I'd rather have another PR for this so we can keep track of each change individually (feel free to open it ;) )

Reply:
There is a drawback when the indices are not saved in a private attribute: if `random_state` is a RandomState instance, the RandomState instance is mutated when it is used ([see the comment of jnothman](https://github.com/scikit-learn/scikit-learn/pull/14012#issuecomment-498998814)), and thus the indices are different when they are computed the second time. However, the indices are identical when using an integer for `random_state`.

The current test does not check this, as the training set has 100 samples (and the subsampling is only used when they are more than 10k samples).

Maybe the best solution is to mention that `random_state` should be an integer when using `warm_start` (and to force it in the validation of the parameters).